sprocs/assessment_instances_points.sql

```diff
@@ -34,81 +29,84 @@ BEGIN
-        -- for Homeworks, drop deleted questions
-        SELECT sum(iq.points), sum(iq.points_in_grading)
-        INTO    total_points,   total_points_in_grading
-        FROM
-            instance_questions AS iq
-            JOIN assessment_questions AS aq ON (aq.id = iq.assessment_question_id)
-        WHERE
-            iq.assessment_instance_id = assessment_instances_points.assessment_instance_id
-            AND aq.deleted_at IS NULL;
-    ELSE
+    IF NOT ((assessment_type = 'Exam') OR (assessment_type = 'Homework')) THEN
         RAISE EXCEPTION 'Unknown assessment_type: %', assessment_type;
     END IF;
 
-    SELECT ai.max_points INTO max_points
-    FROM assessment_instances AS ai
-    WHERE ai.id = assessment_instance_id;
-
-    SELECT ai.score_perc INTO current_score_perc
-    FROM assessment_instances AS ai
-    WHERE ai.id = assessment_instance_id;
-
     -- #########################################################################
-    -- awarded points and score_perc
-
-    -- compute the score in points, maxing out at max_points
-    points := least(total_points, max_points);
-
-    -- compute the score as a percentage, applying credit bonus/limits
-    score_perc := points
-        / (CASE WHEN max_points > 0 THEN max_points ELSE 1 END) * 100;
-    IF credit < 100 THEN
-        score_perc := least(score_perc, credit);
-    ELSIF (credit > 100) AND (points = max_points) THEN
-        score_perc := credit;
-    END IF;
-
-    IF NOT allow_decrease THEN
-        -- no matter what, don't decrease the score_perc
-        score_perc := greatest(score_perc, current_score_perc);
-    END IF;
-
-    -- #########################################################################
-    -- in_grading versions of points and score_perc
-    -- computed by finding max_possible points and score_perc
-    -- and then subtracting off the new values of each quantity
-
-    -- repeat calculation for points_in_grading
-    max_possible_points := least(points + total_points_in_grading, max_points);
-    total_points_in_grading := max_possible_points - points;
-
-    -- compute max achieveable score_perc if all grading points are awarded
-    max_possible_score_perc := max_possible_points
-        / (CASE WHEN max_points > 0 THEN max_points ELSE 1 END) * 100;
-    IF credit < 100 THEN
-        max_possible_score_perc := least(max_possible_score_perc, credit);
-    ELSIF (credit > 100) AND (max_possible_points = max_points) THEN
-        max_possible_score_perc := credit;
-    END IF;
-    
-    IF NOT allow_decrease THEN
-        -- no matter what, don't decrease the achieveable score_perc below new score_perc
-        max_possible_score_perc := greatest(max_possible_score_perc, score_perc);
-    END IF;
-
-    -- compute score_perc_in_grading
-    score_perc_in_grading := max_possible_score_perc - score_perc;
+    -- compute the points by zone
+    RETURN QUERY
+        WITH all_questions AS (
+            SELECT
+                iq.id AS qid,
```

Comment:
Can we call this `iq_id` rather than `qid`? We are already using `qid` to specifically mean the short names of questions (the column `questions.qid`).

Reply:
good point, done.